src/packages/auth/adapters/CAMAuthAdapter.ts

```diff
@@ -94,14 +103,35 @@ async function loginSSO(ssoToken: any): Promise<AuthResponse> {
       };
     }
 
-    const { allowed_roles, default_role } = await getUserRoles(userId, DEFAULT_ROLE, ALLOWED_ROLES);
+    const { default_role, allowed_roles } = mapGroupsToRoles(groupList);
+
+    // if auth group mappings exist, the auth provider and mappings
+    // are the source of truth, so we need to upsert roles in the DB.
+    // We only do this when the user is logging in for the first time
+    // in a session (user cookie DNE) for DB performance reasons,
+    // and to avoid roles changing underneath the user during a session
+    const isNewUserSession = !(userCookie && cookieIsValid(userCookie));
```

Comment:
In the SSO case, the user token could still be valid but the user has to relog due to SSO session expiry

Reply:
Might be a UI bug?

Reply:
I've significantly reworked this section. The conditional logic for upserting roles to the DB now depends on diffing the freshly mapped allowed roles and the existing DB allowed roles for this user.

This entirely removes the "session" concept; users are auth'd and roles are updated exactly when we detect, not just once the first time users log in. This significantly simplifies upstream (i.e. the UI), since we don't have to worry about clearing user tokens there to get a new "session", instead the a Gateway will always provide a JWT that matches the allowed roles in the DB.

I've done some testing, and the UI handles roles changing underneath it quite well. I only see one small UI change (resetting active role cookie to default if it's no longer included in your new allowed roles) that accounts for the edge case of losing a role while using the UI.